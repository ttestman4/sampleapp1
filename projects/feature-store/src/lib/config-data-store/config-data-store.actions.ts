import { Action } from '@ngrx/store';
import { Airport } from './config-data-store.models';

export enum ConfigDataActionTypes {
    // LoadConfig = '[ConfigData] Load Config Data',
    // LoadAirports = '[ConfigData] Load Airports',
    LoadAirportsSuccess = '[ConfigData] Load Airports Success',
    // LoadAirportsFailuer = '[ConfigData] Load Airports Failuer',
}

/**
 * Every action is comprised of at least a type and an optional
 * payload. Expressing actions as classes enables powerful
 * type checking in reducer functions.
 *
 * See Discriminated Unions: https://www.typescriptlang.org/docs/handbook/advanced-types.html#discriminated-unions
 */
// export class LoadConfig implements Action {
//     readonly type = ConfigDataActionTypes.LoadConfig;

//     constructor() { }
// }

// export class LoadAirports implements Action {
//     readonly type = ConfigDataActionTypes.LoadAirports;

//     constructor() { }
// }

export class LoadAirportsSuccess implements Action {
    readonly type = ConfigDataActionTypes.LoadAirportsSuccess;

    constructor(public payload: Airport[]) { }
}

// export class LoadAirportsFailuer implements Action {
//     readonly type = ConfigDataActionTypes.LoadAirportsSuccess;

//     constructor() { }
// }

/**
 * Export a type alias of all actions in this action group
 * so that reducers can easily compose action types
 */
export type ConfigDataActionsUnion =
    // LoadConfig |
    // LoadAirports |
    LoadAirportsSuccess; // |
   // LoadAirportsFailuer;

