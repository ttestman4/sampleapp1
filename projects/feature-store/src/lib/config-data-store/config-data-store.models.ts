export interface Airport {
    code: string;
    name: string;
}

export interface ConfigData {
    airports: Airport[];
}
