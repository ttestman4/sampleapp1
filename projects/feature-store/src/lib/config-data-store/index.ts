export * from './config-data-store.module';
