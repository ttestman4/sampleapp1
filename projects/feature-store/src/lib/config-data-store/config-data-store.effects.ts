import { Injectable } from '@angular/core';
import { Actions, Effect, ofType } from '@ngrx/effects';
import { defer, of } from 'rxjs';
import { catchError, flatMap, map, switchMap } from 'rxjs/operators';
import * as ConfigStoreActions from './config-data-store.actions';
import { Airport } from './config-data-store.models';
import { ConfigDataStoreService } from './config-data-store.service';

let nextConfigDataEffectsId = 1;

@Injectable({
    providedIn: 'root'
})
export class ConfigDataEffects {

    @Effect()
    loadAirports$ = this.actions$.pipe(
        ofType<ConfigStoreActions.LoadAirports>(ConfigStoreActions.ConfigDataActionTypes.LoadAirports),
        switchMap((_action) => {
            return this.configService.getAirports().pipe(
                map((airports: Airport[]) => {
                    return new ConfigStoreActions.LoadAirportsSuccess(airports);
                }),
                catchError(_err => {
                    return of(new ConfigStoreActions.LoadAirportsFailuer());
                }));
        })
    );

    @Effect()
    loadConfigData$ = this.actions$.pipe(
        ofType<ConfigStoreActions.LoadConfig>(ConfigStoreActions.ConfigDataActionTypes.LoadConfig),
        flatMap((_action) => {
            return [new ConfigStoreActions.LoadAirports()];
        })
    );

    // If you want to trigger another action,
    // be careful to add this effect at the end.
    // Should be your last effectyarn build
    @Effect()
    init$ = defer(() => {
        return of(new ConfigStoreActions.LoadConfig());
    });

    constructor(
        private actions$: Actions<ConfigStoreActions.ConfigDataActionsUnion>,
        private configService: ConfigDataStoreService
    ) {
        const id = nextConfigDataEffectsId++;
        if (id > 1) {
            throw new Error(
                'ConfigDataEffects is already loaded. Import it in the AppModule only');
        }
    }
}
