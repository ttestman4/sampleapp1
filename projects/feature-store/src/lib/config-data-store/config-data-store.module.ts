import { CommonModule } from '@angular/common';
import { ModuleWithProviders, NgModule, Optional, SkipSelf } from '@angular/core';
import { StoreModule } from '@ngrx/store';
import { reducer } from './config-data-store.reducer';
@NgModule({
  declarations: [],
  imports: [
    CommonModule,
    /**
     * StoreModule.forFeature is used for composing state
     * from feature modules. These modules can be loaded
     * eagerly or lazily and will be dynamically added to
     * the existing state.
     */
    StoreModule.forFeature('configdata', reducer),
  ]
})
export class ConfigDataStoreModule {
  constructor(@Optional() @SkipSelf() parentModule: ConfigDataStoreModule) {
    if (parentModule) {
      throw new Error(
        'ConfigDataStoreModule is already loaded. Import it in the AppModule only');
    }
  }
  static forRoot(): ModuleWithProviders {
    return {
      ngModule: ConfigDataStoreModule,
      providers: [
      ]
    };
  }
  static forChild(): ModuleWithProviders {
    return {
      ngModule: ConfigDataStoreModule,
      providers: [
      ]
    };
  }
  static forTestReset() {
  }
}
