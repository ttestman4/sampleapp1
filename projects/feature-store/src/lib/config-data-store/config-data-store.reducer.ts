import { createFeatureSelector } from '@ngrx/store';
import { ConfigDataActionsUnion, ConfigDataActionTypes, LoadAirportsSuccess } from './config-data-store.actions';
import { Airport, ConfigData } from './config-data-store.models';

export const featureName = 'configdata';

export interface State extends ConfigData {
    name: string;
}

export const initialState: State = {
    name: 'Config Data Store',
    airports: [],
};

export function reducer(
    state = initialState,
    action: ConfigDataActionsUnion
): State {
    switch (action.type) {
        case ConfigDataActionTypes.LoadAirportsSuccess: {
            const localAction = <LoadAirportsSuccess>action;
            let airports: Airport[];
            if (Array.isArray(localAction.payload)) {
                airports = localAction.payload;
            } else {
                airports = initialState.airports;
            }
            const result: State = {
                ...state,
                airports: airports
            };
            return result;
        }
        default: {
            return state;
        }
    }
}

export const selectConfigData = createFeatureSelector<ConfigData>(featureName);
