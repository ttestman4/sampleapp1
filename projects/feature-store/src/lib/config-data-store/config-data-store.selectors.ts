import { createSelector } from '@ngrx/store';
import { selectConfigData } from './config-data-store.reducer';

export const selectAirports = createSelector(selectConfigData,
    (configData) => configData.airports);
