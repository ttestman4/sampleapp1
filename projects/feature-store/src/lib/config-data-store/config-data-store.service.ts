import { HttpClient } from '@angular/common/http';
import { Injectable } from '@angular/core';
import { Observable } from 'rxjs';
import { map } from 'rxjs/operators';
import { Airport , AirportData} from './config-data-store.models';

let nextConfigDataStoreServiceId = 1;

export function ResetNextConfigDataStoreServiceId() {
  nextConfigDataStoreServiceId = 1;
}


@Injectable({
  providedIn: 'root'
})
export class ConfigDataStoreService {
  airportUrl = 'assets/airports.json';

  getAirports(): Observable<Airport[]> {
    return this.http.get<AirportData>(this.airportUrl).pipe(
      map((response) => {
        return response.data.map((airport) => {
          airport.displayText = airport.city + ' (' + airport.code + ')';
          return airport;
        });
      })
    );
  }

  constructor(private http: HttpClient) {
    const id = nextConfigDataStoreServiceId++;
    if (id > 1) {
      throw new Error(
        'ConfigDataStoreService is already loaded. Import it in the AppModule only');
    }
  }
}
