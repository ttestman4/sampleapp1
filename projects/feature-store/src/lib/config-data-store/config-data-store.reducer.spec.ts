import { LoadAirportsSuccess } from './config-data-store.actions';
import { Airport } from './config-data-store.models';
import { initialState, reducer, State } from './config-data-store.reducer';
describe('Root Reducer', () => {
    it('Unknown action should return the initial state', () => {
        const action = {} as any;

        const result = reducer(initialState, action);

        expect(result).toMatchSnapshot();
    });

    it('Unknown action should return the initial state for undefined initial state input', () => {
        const action = {} as any;

        const result = reducer(undefined, action);

        expect(result).toMatchSnapshot();
    });

    let resultFromPreviousStep: State;
    it('Valid action should return the updated State', () => {
        const airports: Airport[] = [
            { code: 'BOM', name: 'MUMBAI' },
            { code: 'GOI', name: 'GOA' },
        ];
        const action = new LoadAirportsSuccess(airports);

        resultFromPreviousStep = reducer(initialState, action);

        expect(resultFromPreviousStep.airports.length).toEqual(airports.length);
        expect(resultFromPreviousStep.airports).toEqual(airports);
        expect(resultFromPreviousStep).toMatchSnapshot();
    });

    it('Valid action with invalid payload i.e. Non arry should be handled gracefully',
        () => {
            const action = new LoadAirportsSuccess(<Airport[]>9);

            resultFromPreviousStep = reducer(initialState, action);

            expect(resultFromPreviousStep.airports.length).toEqual(initialState.airports.length);
            expect(resultFromPreviousStep.airports).toEqual(initialState.airports);
            expect(resultFromPreviousStep).toMatchSnapshot();
         });
});
