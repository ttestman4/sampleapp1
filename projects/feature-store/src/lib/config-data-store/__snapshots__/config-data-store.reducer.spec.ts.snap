// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Root Reducer Unknown action should return the initial state 1`] = `
Object {
  "airports": Array [],
  "name": "Config Data Store",
}
`;

exports[`Root Reducer Unknown action should return the initial state for undefined initial state input 1`] = `
Object {
  "airports": Array [],
  "name": "Config Data Store",
}
`;

exports[`Root Reducer Valid action should return the updated State 1`] = `
Object {
  "airports": Array [
    Object {
      "code": "BOM",
      "name": "MUMBAI",
    },
    Object {
      "code": "GOI",
      "name": "GOA",
    },
  ],
  "name": "Config Data Store",
}
`;

exports[`Root Reducer Valid action with invalid payload i.e. Non arry should be handled gracefully 1`] = `
Object {
  "airports": Array [],
  "name": "Config Data Store",
}
`;
